### PR TITLE
SPR1-2185: ADD created to pruned products; Add proposalID to vis and …

### PR DESCRIPTION
**Pull request by KgomotsoMonnakgotla**
Add functionality to set pruned products state on solr doc to CREATED. On L0 and L1 products, I've added proposal ID to appear on Solr.

### CODE REVIEW

File: katsdpdata/met_extractors.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 import os
 import subprocess
 import time
+from pudb import set_trace
```
Comment (cchristelis) [inline katsdpdata/met_extractors.py:9]: please remove testing code

File: katsdpdata/met_extractors.py [hunk 2]
```diff
@@ -176,16 +177,19 @@ def _extract_location_from_katdata(self):
                 az, el = t.azel()
                 az, el = katpoint.rad2deg(az), katpoint.rad2deg(el)
                 if -90 <= el <= 90:
-                    self.metadata["ElAz"].append("%f, %f" % (el, katpoint.wrap_angle(az, 360)))
+                    self.metaqqdata["ElAz"].append("%f, %f" % (el, katpoint.wrap_angle(az, 360)))
```
Comment (cchristelis) [inline katsdpdata/met_extractors.py:180]: typo?

File: katsdpdata/met_extractors.py [hunk 3]
```diff
@@ -176,16 +177,19 @@ def _extract_location_from_katdata(self):
...
                 else:
                     self.metadata["ElAz"].append("%f, %f" % ((np.clip(el, -90, 90)), katpoint.wrap_angle(az, 360)))
 
     def _extract_metadata_for_project(self):
+        #set_trace()
```
Comment (cchristelis) [inline katsdpdata/met_extractors.py:185]: please remove

File: katsdpdata/met_extractors.py [hunk 4]
```diff
@@ -176,16 +177,19 @@ def _extract_location_from_katdata(self):
...
         """Populate self.metadata: Grab if available proposal, program block and project id's
         from the observation script arguments."""
         # ProposalId
         if 'proposal_id' in self._katdata.obs_params:
-            self.metadata['ProposalId'] = self._katdata.obs_params['proposal_id']
+            self.metadata['ProposalId'] = self._katdata.obs_params.get('proposal_id',"kgomotso-testing proposal_id field")
```
Comment (cchristelis) [inline katsdpdata/met_extractors.py:190]: more testing code?

File: katsdpdata/met_extractors.py [hunk 5]
```diff
@@ -176,16 +177,19 @@ def _extract_location_from_katdata(self):
...
+            #print(self._katdata.obs_params['proposal_id'])
```
Comment (cchristelis) [inline katsdpdata/met_extractors.py:191]: remove rather than comment out

File: katsdpdata/met_extractors.py [hunk 6]
```diff
@@ -449,6 +453,7 @@ def __init__(self, cbid_stream_rdb_file):
         self.product_type = 'MeerKATTelescopeProduct'
 
     def extract_metadata(self):
+        #set_trace()
```
Comment (cchristelis) [inline katsdpdata/met_extractors.py:456]: rm

File: katsdpdata/met_extractors.py [hunk 7]
```diff
@@ -468,12 +473,16 @@ def extract_metadata(self):
             logger.warning("Metadata already extracted. Set the metadata_extracted attribute to False and run again.")
 
     def _extract_metadata_for_capture_stream(self):
+        #set_trace()
```
Comment (cchristelis) [inline katsdpdata/met_extractors.py:476]: rm

File: katsdpdata/met_extractors.py [hunk 8]
```diff
@@ -468,12 +473,16 @@ def extract_metadata(self):
...
         """Extract CaptureStreamId, CaptureBlockId and StreamId.
         """
         self.metadata['CaptureBlockId'] = self._katdata.source.metadata.attrs['capture_block_id']
         self.metadata['StreamId'] = self._katdata.source.metadata.attrs['stream_name']
         self.metadata['CaptureStreamId'] = self.metadata['CaptureBlockId'] + '_' + self.metadata['StreamId']
         self.metadata['Prefix'] = self._katdata.source.metadata.attrs['capture_block_id']
+        print(self._katdata.source.metadata.attrs)
+        #self.metadata['ProposalID'] = self._katdata.source.metadata.attrs['capture_block_id']
+        self.metadata['ProposalId'] = 'Testing'
```
Comment (cchristelis) [inline katsdpdata/met_extractors.py:485]: testing code? probably possible to remove all of this?

File: katsdpdata/met_extractors.py [hunk 9]
```diff
@@ -502,6 +511,7 @@ def __init__(self, cbid_stream_rdb_file):
         self.product_type = 'MeerKATFlagProduct'
 
     def extract_metadata(self):
+        #set_trace()
```
Comment (cchristelis) [inline katsdpdata/met_extractors.py:514]: rm

File: katsdpdata/met_extractors.py [hunk 10]
```diff
@@ -523,6 +533,7 @@ def _extract_metadata_for_capture_stream(self):
         self.metadata['CaptureStreamId'] = self.metadata['CaptureBlockId'] + '_' + self.metadata['StreamId']
         self.metadata['Prefix'] = self._katdata.source.metadata.attrs['capture_block_id']
 
+
```
Comment (cchristelis) [inline katsdpdata/met_extractors.py:536]: please ensure black/pep-8 compliance

File: katsdpdata/met_handler.py [hunk 11]
```diff
@@ -145,6 +145,7 @@ def __init__(self, *argv, **kwargs):
         super().__init__(*argv, **kwargs)
 
     def create_core_met(self):
+        #set_trace()
```
Comment (cchristelis) [inline katsdpdata/met_handler.py:148]: rm

File: katsdpdata/prod_handler.py [hunk 12]
```diff
@@ -11,6 +11,9 @@
 import shutil
 import time
 from pathlib import PurePath
+from pudb import set_trace
```
Comment (cchristelis) [inline katsdpdata/prod_handler.py:14]: testing code?

File: katsdpdata/prod_handler.py [hunk 13]
```diff
@@ -174,6 +177,7 @@ def __init__(self, product_path):
         self._mh = None
 
     def mh(self, product_type=None):
+        #set_trace()
```
Comment (cchristelis) [inline katsdpdata/prod_handler.py:180]: rm

File: katsdpdata/prod_handler.py [hunk 14]
```diff
@@ -378,6 +382,7 @@ def __init__(self, product_path):
         self.met_handler = MetaDataHandler
         self.product_type = 'MeerKATTelescopeProduct'
 
+
```
Comment (cchristelis) [inline katsdpdata/prod_handler.py:385]: pep8 rm

File: katsdpdata/prod_handler.py [hunk 15]
```diff
@@ -396,6 +401,7 @@ def discover_trawl_files(self):
             self.file_matches = []
 
     def metadata_transfer_complete(self):
+        set_trace()
```
Comment (cchristelis) [inline katsdpdata/prod_handler.py:404]: testing code?

File: katsdpdata/prod_handler.py [hunk 16]
```diff
@@ -483,6 +501,125 @@ def rdb_file_prefix(self):
...
+                        err.filename = rdbs[0]
+                        logger.exception(
+                            "Caught exception while extracting metadata from %s.",
+                            err.filename)
+                        self.set_failed_token(str(err))
+                        # if failed, set a boolean flag to exit the loop.
+                    else:
+                        raise
+
+    def rdb_file_prefix(self):
```
Comment (cchristelis) [inline katsdpdata/prod_handler.py:614]: This is unchanged from the base class, please remove.

File: katsdpdata/prod_handler.py [hunk 17]
```diff
@@ -483,6 +501,125 @@ def rdb_file_prefix(self):
...
+                mh.product_id)
+            err.bucket_name = self.bucket_name()
+            raise err
+        met_original_refs = list(original_refs)
+        met_original_refs.insert(0, os.path.dirname(
+            os.path.commonprefix(original_refs)))
+        met = mh.add_ref_original(met, met_original_refs)
+        mh.add_prod_met(met, pm_extractor.metadata)
+
+    def metadata_when_created(self):
```
Comment (cchristelis) [inline katsdpdata/prod_handler.py:586]: unchanged from base class remove.

**Review by cchristelis — CHANGES_REQUESTED** [review]
Hi @KgomotsoMonnakgotla, this is a good start. I've stopped my review about half way. Please remove all testing code as well as all the sub-classes's methods that didn't change from the base-class. This is unnecessary duplication and looses the focus of the change.

File: katsdpdata/prod_handler.py (hunk 17 above)
Comment (KgomotsoMonnakgotla) [reply katsdpdata/prod_handler.py:586]: If I don't have a return/pass statement in this method for pruned products, this will throw an error. I'm actually overring the method.

Comment (cchristelis) [reply katsdpdata/prod_handler.py:586]: ah, yes, but then the rest of the code should go...

Comment (KgomotsoMonnakgotla) [reply katsdpdata/prod_handler.py:581]: okay,I see what you mean.

File: katsdpdata/prod_handler.py [hunk 18]
```diff
@@ -22,8 +24,8 @@
 logger = logging.getLogger(__name__)
 
 CAPTURE_BLOCK_REGEX = "^[0-9]{10}$"
-CAPTURE_STREAM_L0_REGEX = "^[0-9]{10}[-_].*l0$"
-CAPTURE_STREAM_L1_REGEX = "^[0-9]{10}[-_].*l1-flags$"
+CAPTURE_STREAM_L0_REGEX = "^[0-9]{10}[-_].*l0*$"
```
Comment (ctgschollar) [inline katsdpdata/prod_handler.py:27]: Why did you add this * here?

File: katsdpdata/prod_handler.py [hunk 19]
```diff
@@ -406,6 +408,17 @@ def metadata_transfer_complete(self):
         met = mh.set_product_received(met)
         met_bucket = self.get_bucket_stats()
         mh.add_bucket_stats(met, met_bucket)
+        self.solr = pysolr.Solr(self.solr_url)
+        for product_type in ['MeerKATVisibilityProduct', 'MeerKATFlagProduct']:
+            query = 'CaptureBlockId:{} AND CAS.ProductTypeName:{}'.format(met['id'],product_type)
```
Comment (ctgschollar) [inline katsdpdata/prod_handler.py:413]: You should use met['CaptureBlockId'], not met['id']. The id might not always be a CaptureBlockId.

You could also make your query

```
query = 'CaptureBlockId:{} AND CAS.ProductTypeName:MeerKATVisibilityProduct OR CAS.ProductTypeName:MeerKATFlagProduct'.format(met['CaptureBlockId'])
```

Then you can loop through the results. No point doing 2 queries to solr when you can do one query.

File: katsdpdata/prod_handler.py [hunk 20]
```diff
@@ -406,6 +408,17 @@ def metadata_transfer_complete(self):
...
+            query_dict = {"q": query}
+            results = self.solr.search(**query_dict)
+            if results.hits == 1:
+                doc = results.docs[0]
+                if met.get('ProposalId'):
```
Comment (ctgschollar) [inline katsdpdata/prod_handler.py:418]: I prefer 
```
if 'ProposalId' in met:
```
to 
```
if met.get('ProposalId'):
```
It is just easier to read

File: katsdpdata/prod_handler.py [hunk 21]
```diff
@@ -406,6 +408,17 @@ def metadata_transfer_complete(self):
         met = mh.set_product_received(met)
         met_bucket = self.get_bucket_stats()
         mh.add_bucket_stats(met, met_bucket)
+        self.solr = pysolr.Solr(self.solr_url)
+        for product_type in ['MeerKATVisibilityProduct', 'MeerKATFlagProduct']:
+            query = 'CaptureBlockId:{} AND CAS.ProductTypeName:{}'.format(met['id'],product_type)
+            query_dict = {"q": query}
+            results = self.solr.search(**query_dict)
+            if results.hits == 1:
```
Comment (ctgschollar) [inline katsdpdata/prod_handler.py:416]: What happens if there are no results? Does this code ever get run again?

If this code never gets run again and some vis and flag products are made later, then they will never have the ProposalId set?

File: katsdpdata/prod_handler.py [hunk 22]
```diff
@@ -549,16 +656,20 @@ def __init__(self, trawl_dir):
         capture_block_dirs: list : full path to valid capture block directories
         capture_stream_dirs: list : full path to valid capture stream directories
         """
+        self.solr = pysolr.Solr(solr_url)
         # get full path to capture block dirs
         self.capture_block_dirs = self._list_dir_helper(
             trawl_dir, CAPTURE_BLOCK_REGEX)
+        #print(self.capture_block_dirs, trawl_dir, CAPTURE_BLOCK_REGEX)
```
Comment (ctgschollar) [inline katsdpdata/prod_handler.py:663]: Remove debug prints

**Review by ctgschollar — CHANGES_REQUESTED** [review]
A few changes needed and I am still concerned that you may get to a point where the rdb exists, but not the vis and flag products. In that case we need to set the rdb to an error state or we need to make sure that we keep trying to ingest the rdb until we see the vis and flag products

File: katsdpdata/prod_handler.py (hunk 18 above)
Comment (KgomotsoMonnakgotla) [reply katsdpdata/prod_handler.py:27]: I added the symbol * because incomplete rdb files for continuum products were not considered pruned products.

File: katsdpdata/prod_handler.py (hunk 21 above)
Comment (KgomotsoMonnakgotla) [reply katsdpdata/prod_handler.py:416]: After re-examining the code, I discovered the following.
If  both the  vis  and the flag products dir are empty, and there is no complete token in the rdb directory, the product will be marked as pruned . However, if the complete token is present in the RDB directory and there are no vis and flags directories, the code will move the rdb files to a failed directory.

In other words, this code will be executed if both the vis and flags products are present.

**Review by ctgschollar — APPROVED** [review]
Looks good we need to book a slot to deploy

File: scripts/vis_trawler.py [hunk 23]
```diff
@@ -69,7 +70,7 @@ def main(trawl_dir, boto_dict, solr_url):
 
 
 def trawl(trawl_dir, boto_dict, solr_url):
-    """Main action for trawling a directory for ingesting products
+    """Main action for trawqling a directory for ingesting products
```
Comment (cchristelis) [inline scripts/vis_trawler.py:73]: typo?

**Review by cchristelis — APPROVED** [review]
LGTM
